CONTRIBUTING.md

```diff
@@ -117,7 +117,7 @@ We definitely appreciate pull requests that highlight or reproduce a problem, ev
 
             a. Add new redirect url : <ngrok url>/api/v1/auth/oauth
             b. Click save urls
-            c. Add the following scopes : bot, commands, link:read, link:write, chat:write:bot
+            c. Add the following scopes : bot, commands, link:read, link:write, chat:write, im:read, im:write, channels:read, groups:read, groups:write, npim:write
```

Comment:
What does the rollout process look like on the production Slack app? Will it need to be updated with these scopes whenever this commit is merged, or will it be released later?

If I recall correctly from other Slack apps, you might have to get users to re-install your app with any new scopes you add so they can decide if they want to grant you the new permissions.

Reply:
I can double-check with @kennyshittu, but looking at the CircleCI pipeline, there doesn't seem to be any steps related to deployment. I would assume it gets released later.

When I added the scopes, I received a warning that for a distributed app on the Slack App Directory, we would have to resubmit. You're probably right in that users would have to reinstall. I'm not sure if this is avoidable though.

Reply:
Is there any migration doc from Slack on these deprecations?

Reply:
> Is there any migration doc from Slack on these deprecations?

https://api.slack.com/changelog/2020-01-deprecating-antecedents-to-the-conversations-api

Reply:
According to @kennyshittu, a deploy to Heroku will be triggered when the commit to merge. If that's the case, I think we may want to hold off merging to master for now. I'll need to add the `incoming-webhook` scope to [generate a webhook URL](https://api.slack.com/messaging/webhooks#incoming_webhooks_programmatic) for each workspace that our app is installed in. Since we'll have to resubmit to the Slack App Directory and potentially get users to reinstall the app, we should get all of the new scopes in at once.

Reply:
Do I understand correctly that this changing of scopes is a one time thing for us to do on our side? Not something that people using our app need to change?

Reply:
Deploying to a staging environment is great! Thanks for checking on that @danielpeng2.

I expect the rollout process would look something like:

1. Merge this PR to staging and test there (with a staging Slack app? If it exists?)
2. Update production Slack app config to use new scopes. This will _not_ automatically make user-facing changes
3. Re-submit production Slack app for review by the Slack team
4. Once the Slack app is approved for publish, deploy this code to production + release the Slack app
5. Get production Slack app end-users to re-install the app with the new scopes for the new API calls

If that's the case, then it's possible that the server will need to be able to handle requests from installed apps with the old scopes and the new scopes at the same time until everyone has updated.

**Big Caveat:** It's possible I'm overcomplicating this. Most of my experience here comes from updating an app from OAuth to OAuth 2.0 and from legacy scopes to granular scopes: 
- https://api.slack.com/start/distributing/directory#resubmitting
- https://api.slack.com/authentication/migration#faq_directory_development

That's a bit more involved process than just adding a few more scopes, but I'm _fairly certain_ some of that applies here (where end-users have to manually re-install the app to approve new scopes, they aren't just auto-added when you re-release a live app.)

I'm also happy to chat more synchronously if this is getting confusing!

Reply:
Yes, although users may receive a prompt that they need to reinstall their app.

Reply:
There is a staging Slack app which we can use for testing. As far as I can tell, that rollout process seems right to me.

On the topic of OAuth vs OAuth 2.0, I was just reading up on this. Our app is currently a *classic* Slack bot using the old OAuth, so we'll likely have to perform that migration as well. In November 2021, Slack is deprecating classic bots, so I think we'll need to perform this upgrade before then.

Reply:
It's possible we might have to upgrade sooner. I just saw this line in one of the docs I was trying to reference:

> December 4, 2020 - no updates accepted to App Directory listings using classic bot tokens.

https://api.slack.com/authentication/migration#faq_directory_schedule

That's tomorrow, right?

Reply:
Ah you're right. Our current app as-is will work fine until November 2021, but we can't update it with new scopes after tomorrow.

In your experience with doing the OAuth migration, how long does it take? In the interest of time I may work on getting the webhook functionality working in staging first, before we handle any submissions to the App Directory.

Reply:
The actual scope migration didn't take very long—I wrote out a list of all the legacy scopes we were using and how, then looked up the granular scopes we needed to request instead. The OAuth migration was a bit trickier, in terms of making sure people had the right version/scope to use the appropriate methods and functions.

This doc is really helpful to think through all the steps: https://api.slack.com/authentication/migration

You might also need to budget a few workdays for the Slack team to actually test and approve your app changes once you're ready to re-submit, as a heads up.

I would say most of this isn't _complicated_ as much as it's _tedious_ so defer to your best judgment about whether you'd rather push on webhooks first or get this in order and then do the feature work!

Reply:
Thanks for the input. I think pushing on the webhook functionality would be better first.

On the bright side, I discovered the new Slack app I created for testing was a *new* Slack app. I created a new *classic* Slack app, and it works without adding any new permissions. So we don't have to include those in this PR, we can do that when we actually do the OAuth migration later.

Reply:
Awesome! That sounds like it will simplify things a bit for now, and having both kinds of apps will be helpful for testing I bet. Nice find.